### ISSUE TITLE
multi gpu + multi tower on slurm cluster

### ISSUE THREAD

**Issue by ahundt** [ppwwyyxx/tensorpack#193]
I'm trying to integrate a couple of different components and was hoping to seek advice. TensorPack looks like it has some good [multigpu scripts](https://github.com/ppwwyyxx/tensorpack/blob/master/tensorpack/train/multigpu.py) I'd like to employ for both image labeling and semantic segmentation training.

I have the following:

 - [tensorflow_slurm_manager](https://github.com/jhollowayj/tensorflow_slurm_manager) to get the [tf.train.ClusterSpec](https://www.tensorflow.org/api_docs/python/tf/train/ClusterSpec) once things have started up
 - [slurm cluster with gpus](https://slurm.schedmd.com/)
     - [relevant discussion](https://github.com/jhollowayj/tensorflow_slurm_manager/issues/1) and [exampe](https://github.com/jhollowayj/distributed_tensorflow/blob/slurm_distributed/main.py#L63-71)
 - a couple keras models to train, being discussed in https://github.com/ppwwyyxx/tensorpack/issues/160

A slurm script is just some comments at the top of a shell script that specifies the number of machines, gpus etc and gets lists of what was allocated then runs the programs you wanted later in the shell script.

Basically, examples and usage of AsyncMultiGPUTrainer seem a bit opaque at the moment, though I do see one in the [A3C code](https://github.com/ppwwyyxx/tensorpack/blob/master/examples/A3C-Gym/train-atari.py#L273), and not much that works with the standard TensorFlow ClusterSpec or a multi tower example. 

With all that I have a few questions:

 - Can a multi tower example be created, and if it makes sense to integrate it with a `ClusterSpec`?
 - Will there be any limitations I should be aware of with the TensorPack APIs when training with segmentation data?

**Comment by ppwwyyxx** [comment]
Do you just want multiGPU? Distributed training are not supported at the moment.

If just multiGPU, is slurm anywhere different from a normal machine with multiple GPUs? I know nothing about slurm so I'm not sure what's the issue.

examples/HED is a segmentation example. Nothing really different from other tasks.

**Comment by ahundt** [comment]
Oh I see when I was reading the code I misunderstood tower to mean a cluster with gpus on separate machines. Slurm is like Kubernetes, it is for running in a datacenter or supercomputer with many physical machines, each of which can have different numbers of gpus. Basically you add some specially formatted comments to a shell script requesting CPUs, GPUs, and memory from a cluster then it will run the shell script as requested with environment variables filled out detailing what specific machines you are running on.

Don't worry if multiple machines isn't something tensorpack supports. I wasn't sure, that's why I asked!

**Comment by ppwwyyxx** [comment]
MultiGPU on a single machine should run properly then. Are there unresolved questions?

**Comment by ahundt** [comment]
Nope, thanks!
